### ISSUE TITLE
uri wildcard query

### ISSUE THREAD

**Issue by judell** [hypothesis/product-backlog#30]
## Problem

Every publisher who embeds H on a site (including blog-based sites, esp. WordPress) wants to be able to monitor annotation activity across the site.

At the moment we only support a per-URL API query, so for example the notifier that watches the H site and blog keeps a list of them and queries each individually.

That's untenable, we have many requests for a sitewide query. 

## Possible solution

This was harder before the postgres migration, I believe it is much easier now.  For example:

```
postgres=# select target_uri from annotation where target_uri LIKE '%jonudell.net%';

                  target_uri
-----------------------------------------------
 http://jonudell.net/h/
 http://jonudell.net/h/accessibility/test.html
 http://jonudell.net/h/rhetoric.html
(3 rows)
```

An API facet that does this will be immediately useful to publishers who have been hoping to monitor H activity on their sites using RSS. And while that only works for Public annotations,  because RSS doesn't deal well with authentication, we also have examples -- https://github.com/judell/h_export, which can be used as a widget, e.g. as here, https://hypothes.is/letters-to-the-next-president/, and https://github.com/judell/h_notify -- that publishers can build on until we bake in more robust notification.

Bottom line: A (seemingly) easy thing to do for a large benefit.

**Comment by judell** [comment]
Straw man proposal: https://github.com/hypothesis/h/compare/master...site-query, http://jonudell.net/h/sitequery.mp4

**Comment by malcook** [comment]
Such capability could also be used to search hierarchically organized groups, as suggested in  #401

**Comment by judell** [comment]
See https://github.com/hypothesis/product-backlog/issues/235#issuecomment-357459601 !!!

**Comment by klemay** [comment]
Request from a publisher [in our forum](https://groups.google.com/d/msg/hypothesis-forum/3kBgZAVpvhg/UMnB0DLJCAAJ):
>Hi, is there a way a publisher can view annotation statistics across their entire site / domain? For example, I would like to know how many annotations where made in a particular day and the URLs they were created within. Right now I have over 15K URLs that have the hypothesis embed code so I would like to know which URLs have received annotations. Thanks.

And another request [via Zendesk](https://hypothesis.zendesk.com/agent/tickets/2598).

**Comment by hmstepanek** [comment]
This is more of a feature edition than a refactor of an existing feature and I think we should wait on adding features until after GDPR. That being said, I see the importance of this and this isn't the first time this has been requested (as @klemay has pointed out) so it seems like it would be a good thing to add to the queue of work. Certainly since we will have been digging in that part of the code already I think after GDPR would be a good time to add this. Without doing a bunch of research right now I think we would easily be able to achieve this by adding a regex query filter in elasticsearch. I think the front end would be ok as you can type and search for tags that don't exist in the drop down. I'll let @ajpeddakotla schedule it accordingly.

**Comment by robertknight** [comment]
I'm agreed with Hannah. This is clearly a much needed capability but we need to concentrate on doing the ES upgrade/tests first before adding new features. Once we've got that out of the way it should be much easier to add something like this.

**Comment by judell** [comment]
It already was pretty easy to add, the (maybe) harder thing was to suss out the performance implications of wildcarding. But yeah, now that it seems we're not only changing the ES engine substantially, but also the wrapper we use to talk to it, this should wait. Hopefully, though, it can serve as a reminder that wildcarding is wanted, in case that might affect any implementation decisions.

**Comment by robertknight** [comment]
We have some limited support for URL-based searches using the `uri.parts` field: https://github.com/hypothesis/client/issues/739#issuecomment-396281964

**Comment by klemay** [comment]
Another user request: https://hypothesis.zendesk.com/agent/tickets/2644

> Hi,
> I have embedded hypothesis here: [redacted]
It's a site with more than one page. Is there a way to have a list of all annotation of the entire website?
> 
> Is there some jolly character and create a search like url:http:/mybwebsite.com/*?

**Comment by klemay** [comment]
https://twitter.com/lakens/status/1026811830822465541
"Is it possible in the search or in the rss feed to use a wildcard? I would like to search url:https://psyarxiv.com/ * and get ANY comment made on ANY PsyArxiv preprint. Would be SUPER useful - can it be done?"

**Comment by ajpeddakotla** [comment]
@hmstepanek just to make sure, I've added some acceptance criteria, since this card hasn't been updated in a while: 

- [ ] Query should support true paths and wildcards (ex: nytimes.com, nytimes.com/politics, nytimes/politics/*)
- [ ] Should be able to query from the API
- [ ] Should be able to query from activity pages using the same `url:` facet we already use

**Comment by hmstepanek** [comment]
^ @ajpeddakotla @robertknight I'm nervous about that last point. My concern is any url that currently has a * in it would now not behave the same. I doubt we have many of those but there is that possibility and there's no way to distinguish the difference between a part of the url * and a wildcard *. I'm fine with either decision, I just want to raise this as a potential issue before we merge it with url/uri. Thoughts on this?

**Comment by dwhly** [comment]
> My concern is any url that currently has a * in it would now not behave the same.

Since I was the one that proposed this....

Had to go refresh on [RFC3986](http://www.ietf.org/rfc/rfc3986.txt) to see which chars were qualified as reserved. Surprise.  Chances are probably slim that * gets used much, but we should probably implement best practices, eh?

So is the suggested implementation that all URLs are treated as [partial matches](https://www.elastic.co/guide/en/elasticsearch/guide/current/partial-matching.html)?  

The implication I suppose is that you couldn't search for url:nytimes.com without also getting every partial match.  Returning only the annotations on the nytimes home page would not be possible?

That is of course the benefit of explicitly using a wildcard char.

How do others solve this elegantly....

**Comment by hmstepanek** [comment]
The way I imagined it was if you wanted to search for url:nytimes.com it would only return urls that exactly match nytimes.com. If you instead search for url:`*nytimes.com*` it would return as you suggested above, all matches.

**Comment by dwhly** [comment]
I suppose one approach would be to only interpret "*" at the very end of a URL search as a wildcard (as opposed to URLs with "*" embedded).  Yes, we wouldn't be able to handle wildcards in the middle of strings-- but perhaps an acceptable sacrifice?  We'd find everything with "*" actually at the end of the URL.

**Comment by dwhly** [comment]
I thought:

> My concern is any url that currently has a * in it would now not behave the same. 

Meant that you were concerned about interpreting "*" as a wildcard character vs an actually valid URL char.  The way you've just stated it is what I was expecting.

> If you instead search for `url:*nytimes.com*` it would return as you suggested above, all matches.

**Comment by hmstepanek** [comment]
You interpreted that correctly, I just wanted to clarify what I was thinking for user input as well since it seems like we had different thoughts surrounding that. url:nytimes.com would not trigger a wildcard. A wildcard is only triggered if there's a `*` in the value (at least that was what I had in mind). Sounds like we're on the same page there so sorry for any confusion. 

The option of only interpreting `*`'s at the edges of the world as wildcards is certainly an potential solution, but I feel like it's sacrificing a fair amount of functionality for a little problem. 

Another option I guess would be to use different filters when searching in our api vs searching on activity pages. So on activity pages have it behave as a wildcard if there are `*`'s and ignore all cases were there may be `*`'s in the original url, but in our /api/search break it out into a separate keyword so instead of url:`*nytimes.com*` you would use something like url_wildcard:`*nytimes.com*`.  That way it doesn't affect displaying annotations on the sidebar which in some cases queries the api and needs a specific url (aka no wildcards allowed).

**Comment by dwhly** [comment]
> A wildcard is only triggered if there's a * in the value (at least that was what I had in mind). 

Me too.

> So on activity pages have it behave as a wildcard if there are *'s and ignore all cases were there may be *'s in the original url

Interestingly, the "*" as a wildcard would of course match actual *s in the URLs, so it would return those too presumably.

> The option of only interpreting *'s at the edges of the world as wildcards is certainly an potential solution, but I feel like it's sacrificing a fair amount of functionality for a little problem.

By sacrifice, I assume you mean it would exclude the potential for searching with *'s inside strings?  For instance, one could imagine searching for all annotations on pdfs at arxiv.org by searching for `url:arxiv.org*pdf`  I can see that this would definitely be useful.

This is related to a question I had actually, which was whether there are any restrictions we will need to impose in order to maintain performance (and not drag down our ES instance).  For instance, will we allow leading *s, etc.  I could imagine that these would be very expensive queries.

> but in our /api/search break it out into a separate keyword so instead of `url:*nytimes.com*` you would use something like `url_wildcard:*nytimes.com*`. That way it doesn't affect displaying annotations on the sidebar which in some cases queries the api and needs a specific url (aka no wildcards allowed).

That seems like a very practical solution.

One more Q:  will `url:nytimes.com*` return all httpx://nytimes.com/* or will we need to query by protocol?  Not sure how it works now.

**Comment by hmstepanek** [comment]
You have to query with the https or http in front but it will be converted to httpx before searching so it would find both https's and http's.

**Comment by hmstepanek** [comment]
Did a search of urls with `*` turns out there are unexpectedly a lot of them (~2500). I also discussed with @robertknight  and sounds like the best approach would prob be the one I mentioned above (using different filters on activity pages than on the api). Also @robertknight pointed out there is a performance issue with adding `*` to the front of the urls so if it's ok with everyone I'm going to disallow that but `*`'s anywhere else within the url are permitted.

**Comment by dwhly** [comment]
> You have to query with the https or http in front but it will be converted to httpx before searching so it would find both https's and http's.

Maybe we can add it if it's not there?  

> Also @robertknight pointed out there is a performance issue with adding * to the front of the urls so if it's ok with everyone I'm going to disallow that but *'s anywhere else within the url are permitted.

Cool.  Perhaps we can modify the activity pages help text to something like the following:

"url: see all annotations for a given url, path or domain. interior and trailing * wildcards permitted.

(Or do we also support `?`)

**Comment by hmstepanek** [comment]
I think prepending httpx to the beginning opens up a whole can of worms I don't want to get into but  adding a mention about wildcards on activity pages for users is quite do-able. It does also support ?.

**Comment by dwhly** [comment]
> I think prepending httpx to the beginning opens up a whole can of worms I don't want to get into

For activity pages does that mean people will need to know to enter "http://nytimes.com" vs "nytimes.com"?

**Comment by hmstepanek** [comment]
Yes.

**Comment by robertknight** [comment]
@hmstepanek - What is there left that you currently plan to do for this card?

**Comment by aborruso** [comment]
Hi, is it already possible to test wildcard query?

Thank you

**Comment by ajpeddakotla** [comment]
@aborruso through the API, yes. 
Look here for the documentation: https://h.readthedocs.io/en/latest/api-reference/#operation/search. 
<img width="953" alt="screen shot 2018-09-24 at 8 43 16 am" src="https://user-images.githubusercontent.com/7463523/45955651-1752a980-bfd6-11e8-85b2-4ee8d5461e64.png">

Please note that this feature is experimental and is subject to change. If testing goes well on this, we will be releasing this ability to our search page as well, which is here: https://hypothes.is/search

**Comment by malcook** [comment]
This is great.  Hoping,  I tried wildcard to work with RSS tag search, but found it not to work.  Might this capability find a home there too?

**Comment by judell** [comment]
> prepending httpx to the beginning opens up a whole can of worms

Can you say a bit more about that? Because not requiring the user to type an unnecessary 7 or 8 characters would be a major usability win.

**Comment by hmstepanek** [comment]
It's easily done for urls like http://example.com/foo. It's not easily done for urls like server:port/foo. The way to do this would be to pass it into urlparse and if it's missing a scheme (aka http/https) add it but for urls of the second class urlparse thinks the scheme is server. And you might think just append it to everything that doesn't start with http but what about doi urls or urn urls? There isn't a consistent way to do this for all urls which is why it's difficult. It could be we just decide we'll do it for example.com urls and that's fine but smashing that change in with wildcard doesn't make sense. They are two different changes. Dan has opened an issue here: https://github.com/hypothesis/product-backlog/issues/796.

**Comment by robertknight** [comment]
Closing this as done. Any follow-up concerns should be raised in new issues.
